Add options to `decode()` to aid detection.

crankyoldgit opened the pull request:
* All decoding protocols updated to use a starting offset when decoding.
* `irrecv::decode()` now has two additional optional arguments.
  - `max_skip`: Skip over entries at the start of a capture to aggressively look for protocols to decode.
     Warning: Very CPU expensive!
  - `noise_filter`: Try to remove entries from the raw data that are smaller than this value.
     Danger: This will cook the raw data and potentially break some protocol decoders.

* Unit tests updated to use starting offset.
* New unit tests to confirm the new options work as expected.

Fixes #1042

crankyoldgit committed:
Add options to `decode()` to aid detection.
* All decoding protocols updated to use a starting offset when decoding.
* `irrecv::decode()` now has two additional optional arguments.
  - `max_skip`: Skip over entries at the start of a capture to aggressively look for protocols to decode.
     Warning: Very CPU expensive!
  - `noise_filter`: Try to remove entries from the raw data that are smaller than this value.
     Danger: This will cook the raw data and potentially break some protocol decoders.

* Unit tests updated to use starting offset.
* New unit tests to confirm the new options work as expected.

Fixes #1042

crankyoldgit commented:
Code Reviewers,
Apologies in advance, this is a monster PR to review. (Imagine writing, testing, & debugging it!)

FYI @egueli
Can you please test this out extensively?

egueli commented on src/IRrecv.cpp:
```diff
@@ -307,6 +307,35 @@ void IRrecv::setTolerance(const uint8_t percent) {
 // Get the base tolerance percentage for matching incoming IR messages.
 uint8_t IRrecv::getTolerance(void) { return _tolerance; }
 
+#if ENABLE_HIGHPASS_FILTER_OPTION
```
Since we're only allowing high-period, low-_frequency_ pulses I'd suggest to change to
```suggestion
#if ENABLE_LOWPASS_FILTER_OPTION
```
or
```suggestion
#if ENABLE_GLITCH_FILTER_OPTION
```

crankyoldgit replied:
Yes, you are correct. I used a poor name. I've changed it to a more generic `ENABLE_NOISE_FILTER_OPTION` and altered the procedure name as well.

crankyoldgit committed:
Change name of compile-time option for noise filter.

crankyoldgit committed:
Change the name of the proceedure too.

crankyoldgit committed:
Improve detection of NEC repeat codes.

Ref:
  
https://github.com/crankyoldgit/IRremoteESP8266/issues/1042#issuecomment-583874095
  
https://github.com/crankyoldgit/IRremoteESP8266/issues/1042#issuecomment-583874095

crankyoldgit committed:
Improve noise filtering.
* Use a `uint16_t` for the noise floor to allow values large than 255.
* Address an off-by-one issue which allows removal of noise at the end 
of a sample.
* Add safe-guards in case we are near the end of the capture buffer.
* Tweak NEC repeat detection to make it more strict. i.e. Require a 
trailing gap.
  - @egueli may not like this. ;-)
* Add more Unit Tests based on real-world data/conditions.

For #1042

crankyoldgit committed:
Oops. Missed the unit tests it seems.

egueli commented on src/IRrecv.cpp:
```diff
@@ -307,6 +307,37 @@ void IRrecv::setTolerance(const uint8_t percent) {
+    uint16_t curr = results->rawbuf[offset];
+    uint16_t next = results->rawbuf[offset + 1];
+    uint16_t addition = curr + next;
+    if (curr < kTickFloor) {  // Is it to short?
```
```suggestion
    if (curr < kTickFloor) {  // Is it too short?
```

crankyoldgit replied:
Thanks.

crankyoldgit committed:
Fix a typo

Co-Authored-By: Enrico Gueli <enrico.gueli@gmail.com>

crankyoldgit commented:
Confirmed working by @egueli in: https://github.com/crankyoldgit/IRremoteESP8266/issues/1042#issuecomment-583964609

crankyoldgit commented:
Merging after 7 days and no feedback.

jamsinclair commented:
Quite the change and out of my depth of understanding 😅 . On a very superficial level, looks good 👌

crankyoldgit commented:
> Quite the change and out of my depth of understanding 😅 . On a very superficial level, looks good 👌

That's totally understandable. Code reviews don't always need to fully understand it. They are just a sanity check that someone hasn't missed something simple or made an obvious mistake. i.e. It's not your fault if you miss something etc. You're just a second set of eyes that _might_ catch something or suggest a better way etc.

Ref #115 for what I'm looking for in a code review. I should probably put that somewhere on the wiki. ...
